wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1064 @@
+            output[i] += (hex[i * 2 + 1] - '0');
+        else {
+            WOLFSSL_MSG("hex conversion: syntax error");
+            return -1;
+        }
+    }
+    return (int)sz;
+}
+
+static int bytes_to_hex(const unsigned char *bytes, char *hex, unsigned long sz)
+{
+    word32 i;
+    for (i = 0; i < sz; i++) {
+        unsigned char val;
+        val = (bytes[i] >> 4) & 0x0F;
+        if (val > 9)
+            hex[2 * i] = 'A' + (val - 10);
+        else
+            hex[2 * i] = '0' + val;
+        val = (bytes[i] & 0x0F);
+        if (val > 9)
+            hex[2 * i + 1] = 'A' + (val - 10);
+        else
+            hex[2 * i + 1] = '0' + val;
+    }
+    return (int)(2 * sz);
+}
+
+static int expect_tok(const char *cmd, int size, const char *tok, char **repl)
+{
+    int ret;
+    char *r_found = NULL;
+    if (repl)
+        *repl = NULL;
+
+    if (cmd)
+        csim_write(cmd, size);
+    do {
+        ret = csim_read(csim_read_buf, MAXBUF);
+        if (tok && (ret > 0) && !r_found) {
+            r_found = strstr(csim_read_buf, tok);
+            if (r_found)
+                strncpy(parser_line, r_found, MAXBUF / 2 - 1);
+        }
+    } while (ret > 0);
+    if (repl && r_found) {
+        *repl = parser_line + strlen(tok);
+    }
+    if (r_found)
+        return (int)strlen(*repl);
+    else
+        return 0;
+}
+
+static int expect_ok(const char *cmd, int size)
+{
+    return expect_tok(cmd, size, "OK", NULL);
+}
+
+/* Utilities
+ */
+
+/* Conversion utility from hex string format to bytes */
+static int hexbuffer_conv(char *hex_str, unsigned char *output, unsigned long sz)
+{
+    int ret;
+    if (strlen(hex_str) != (2 * sz)) {
+        return -1;
+    }
+    ret = hex_to_bytes(hex_str, output, sz);
+    if (ret != (int)sz)
+        return ret;
+    return (int)sz;
+}
+
+/* Search a TLV by tag in a buffer */
+static char *search_tlv(const char *haystack, int size, uint8_t tag)
+{
+    int i = 0;
+    uint8_t t;
+    uint8_t l;
+    while (i < size) {
+        hex_to_bytes(&haystack[i], &t, 1);
+        hex_to_bytes(&haystack[i + 2], &l, 1);
+        if (tag == t) {
+            return (char *)&haystack[i];
+        }
+        if (l + i > size)
+            return NULL;
+        i += 4 +  2 * l;
+    }
+    return NULL;
+}
+
+static int iotsafe_cmd_start(char *cmd, byte class, byte ins, byte p1, byte p2)
+{
+    byte lc = 0;
+    char *out;
+    memset(cmd, 0, IOTSAFE_CMDSIZE_MAX);
+    strcpy(cmd, "AT+CSIM= 10,\"");
+    out = cmd + AT_CSIM_CMD_SIZE;
+    bytes_to_hex(&class, out, 1);
+    bytes_to_hex(&ins, out + 2, 1);
+    bytes_to_hex(&p1, out + 4, 1);
+    bytes_to_hex(&p2, out + 6, 1);
+    bytes_to_hex(&lc, out + 8, 1);
+    out[10] = 0;
+    return (int)strlen(cmd);
+}
+
+static int iotsafe_cmd_add_tlv16(char *cmd, byte tag, uint16_t len, const byte *val)
+{
+    byte cur_csim_len;
+    byte cur_lc;
+    char *out;
+    int ret;
+    byte *len_b = (byte *)&len;
+
+    if (strlen(cmd) < AT_CSIM_CMD_SIZE)
```

Comment:
Should we be using our macros `XSTRLEN`?

Reply:
ouch, seems like I had missed many of these. Renamed to XSTRLEN/XSTRNCPY/XMEMSET/etc. everywhere now.